runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java

```diff
@@ -117,12 +118,40 @@ public void initPassiveSegment(int segmentId, int dest, int offset, int size) {
         write(dest, segment.data(), offset, size);
     }
 
+    public void writeString(int offset, String data, Charset charSet) {
+        write(offset, data.getBytes(charSet));
+    }
+
     public void writeString(int offset, String data) {
-        write(offset, data.getBytes(StandardCharsets.UTF_8));
+        writeString(offset, data, StandardCharsets.UTF_8);
     }
 
     public String readString(int addr, int len) {
-        return new String(readBytes(addr, len), StandardCharsets.UTF_8);
+        return readString(addr, len, StandardCharsets.UTF_8);
+    }
+
+    public String readString(int addr, int len, Charset charSet) {
```

Comment:
was kind of going back and forth on it. On the one hand, we could let the caller handle some of this charset stuff. But then we'd just be exposing a bytes interface which i think is a little too low level. Granted, maybe we'll take this stuff out eventually and put in some kind of bindings package.